identity-iota/src/credential/validated_data.rs

```diff
@@ -0,0 +1,111 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use self::document_state::Active;
+use self::document_state::Deactivated;
+use serde::Serialize;
+use serde::ser::SerializeStruct;
+use std::collections::BTreeMap;
+use std::marker::PhantomData;
+
+use super::deficiencies::CredentialDeficiencySet;
+
+use identity_core::common::Object;
+
+use identity_credential::credential::Credential;
+use identity_credential::presentation::Presentation;
+
+use crate::did::IotaDID;
+use crate::document::ResolvedIotaDocument;
+
+pub(super) mod document_state {
+  // Used to parameterise whether a resolved document is active.
+  pub trait Sealed {}
+  #[derive(Clone, Copy, Hash, PartialEq, Eq, Debug)]
+  pub struct Active;
+  #[derive(Clone, Copy, Hash, PartialEq, Eq, Debug)]
+  pub struct Deactivated;
+  impl Sealed for Active {}
+  impl Sealed for Deactivated {}
```

Comment:
Not sure I see the advantage of encoding the state of the document in the type rather than in a `bool`. What is the benefit of this approach?

Reply:
It provides more type safety as it is now impossible to use them in the same context and this is also more backward compatible if we want to say make `Deactivated`  unrepresentable, as the type itself although public is not reachable. I actually started using const generics here, so it was `DocumentValidation<true>` and `DocumentValidation<false>` but it is less flexible in terms of backwards compatibility so I dropped that idea.

Reply:
Fair enough, more type safety is nice to have in this case.